pkg/tsdb/parca/query.go

```diff
@@ -123,54 +128,161 @@ type CustomMeta struct {
-		frame := treeToNestedSetDataFrame(flameResponse.Flamegraph)
-		frame.Meta = &data.FrameMeta{PreferredVisualization: "flamegraph"}
-		return frame
+	if flameResponse, ok := resp.Msg.Report.(*v1alpha1.QueryResponse_FlamegraphArrow); ok {
+		return arrowToNestedSetDataFrame(flameResponse.FlamegraphArrow)
 	} else {
-		panic("unknown report type returned from query")
+		// TODO: Can we be nicer about signaling users to update to have the latest APIs?
+		panic("unknown report type returned from query. update parca?")
 	}
 }
 
-// treeToNestedSetDataFrame walks the tree depth first and adds items into the dataframe. This is a nested set format
-// where by ordering the items in depth first order and knowing the level/depth of each item we can recreate the
-// parent - child relationship without explicitly needing parent/child column and we can later just iterate over the
-// dataFrame to again basically walking depth first over the tree/profile.
-func treeToNestedSetDataFrame(tree *v1alpha1.Flamegraph) *data.Frame {
+func seriesToDataFrame(seriesResp *connect.Response[v1alpha1.QueryRangeResponse], profileTypeID string) []*data.Frame {
+	frames := make([]*data.Frame, 0, len(seriesResp.Msg.Series))
+
+	for _, series := range seriesResp.Msg.Series {
+		frame := data.NewFrame("series")
+		frame.Meta = &data.FrameMeta{PreferredVisualization: "graph"}
+		frames = append(frames, frame)
+
+		fields := data.Fields{}
+		timeField := data.NewField("time", nil, []time.Time{})
+		fields = append(fields, timeField)
+
+		labels := data.Labels{}
+		for _, label := range series.Labelset.Labels {
+			labels[label.Name] = label.Value
+		}
+
+		valueField := data.NewField(strings.Split(profileTypeID, ":")[1], labels, []int64{})
+
+		for _, sample := range series.Samples {
+			timeField.Append(sample.Timestamp.AsTime())
+			valueField.Append(sample.Value)
+		}
+
+		fields = append(fields, valueField)
+		frame.Fields = fields
+	}
+
+	return frames
+}
+
+func arrowToNestedSetDataFrame(flamegraph *v1alpha1.FlamegraphArrow) *data.Frame {
 	frame := data.NewFrame("response")
+	frame.Meta = &data.FrameMeta{PreferredVisualization: "flamegraph"}
 
 	levelField := data.NewField("level", nil, []int64{})
 	valueField := data.NewField("value", nil, []int64{})
-	valueField.Config = &data.FieldConfig{Unit: normalizeUnit(tree.Unit)}
+	valueField.Config = &data.FieldConfig{Unit: normalizeUnit(flamegraph.Unit)}
 	selfField := data.NewField("self", nil, []int64{})
-	selfField.Config = &data.FieldConfig{Unit: normalizeUnit(tree.Unit)}
+	selfField.Config = &data.FieldConfig{Unit: normalizeUnit(flamegraph.Unit)}
 	labelField := data.NewField("label", nil, []string{})
 	frame.Fields = data.Fields{levelField, valueField, selfField, labelField}
 
-	walkTree(tree.Root, func(level int64, value int64, name string, self int64) {
+	arrowReader, err := ipc.NewReader(bytes.NewBuffer(flamegraph.GetRecord()))
+	if err != nil {
+		// TODO: Handle properly?
+		return nil
```

Comment:
Can we just return it? I mean before we didn't return `(dataFrame, error)` response but mainly because there was nothing in the processing that would return error that we couldn't handle in place, here though seems like more things can go wrong so I would return it and pass into the `response.Error` in the `query()` function.

Reply:
Yes, happy to return errors. I just thought it was best practice not to return errors here in these plugins.

Reply:
I guess we are not always super consistent with this but errors should be returned upstream and let either front-end or some middleware deal with it, if they cannot be recovered in the plugin.

Reply:
Fixed.